crates/sc-proof-of-time/src/source/gossip.rs

```diff
@@ -324,15 +323,25 @@ where
     /// well as produce next proof if it was already received out of order before
     async fn handle_next_slot_input(&mut self, next_slot_input: PotNextSlotInput) {
         let mut old_proofs = HashMap::<GossipProof, Vec<PeerId>>::new();
-        for (sender, proofs) in &mut self.gossip_cache {
-            proofs.retain(|proof| {
-                if proof.slot > next_slot_input.slot {
-                    true
-                } else {
-                    old_proofs.entry(*proof).or_default().push(*sender);
-                    false
+
+        let changes: HashMap<_, _> = self
+            .gossip_cache
+            .iter()
+            .map(|(sender, proofs)| {
+                let mut retained_proofs = VecDeque::new();
+                for proof in proofs {
+                    if proof.slot > next_slot_input.slot {
+                        retained_proofs.push_back(*proof);
+                    } else {
+                        old_proofs.entry(*proof).or_default().push(*sender);
+                    }
                 }
-            });
+                (*sender, retained_proofs)
+            })
+            .collect();
```

Comment:
actually there is `.retain() `API on `VecDeque<GossipProof>` but there is no `iter_mut` in LruMap, so I cannot mutate the proofs while iterating(if my understanding is correct) . I was trying to avoid this : 

`
 for (sender, proofs) in self.gossip_cache.iter() {
        for proof in proofs {
            if proof.slot <= next_slot_input.slot {
                old_proofs.entry(*proof).or_default().push(*sender);
            }
        }
  }
        
  // this is n³ 
  for (proof, senders) in old_proofs.iter() {
      for sender in senders {
          if let Some(proofs) = self.gossip_cache.get(sender) {
              proofs.retain(|p| p != proof);
              if proofs.is_empty() {
                  self.gossip_cache.remove(sender);
              }
          }
      }
  }
`


I am also okay to implement iter_mut on the LruMap if you think that is the better approach.

Reply:
The absolute numbers should be small there so it is not a lot of iterations. Suggested code is what I expected, but if you can implement mutable iteration that would be even better for sure. Basically old code would not need to be changed.